### PR TITLE
bpo-27575: port set intersection logic into dictview intersection

### PR DESCRIPTION
**Pull request by fgregg**
This turns a patch contributed by @dsu1995 on bpo into a github PR.

The patch ports the logic of set intersection into the intersection of dictviews. It will address performance issues when taking the intersection of large dictviews.

https://bugs.python.org/issue27575

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


When your account is ready, please add a comment in this pull request
and a Python core developer will remove the `CLA not signed` label
to make the bot check again.

Thanks again for your contribution, we look forward to reviewing it!

**Comment by fgregg** [comment]
@dsu1995, have you signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA)? I just signed, and will check back tomorrow to make sure that this status is reflected on bpo.

File: Lib/test/test_dictviews.py [hunk 1]
```diff
@@ -92,13 +92,19 @@ def test_keys_set_operations(self):
...
         self.assertEqual(d1.keys() & d2.keys(), {'b'})
         self.assertEqual(d1.keys() & d3.keys(), set())
         self.assertEqual(d1.keys() & set(d1.keys()), {'a', 'b'})
         self.assertEqual(d1.keys() & set(d2.keys()), {'b'})
         self.assertEqual(d1.keys() & set(d3.keys()), set())
         self.assertEqual(d1.keys() & tuple(d1.keys()), {'a', 'b'})
+        self.assertEqual(d3.keys() & d4.keys(), {'d'})
+        self.assertEqual(d4.keys() & d3.keys(), {'d'})
+        self.assertIsInstance(d4.keys() & frozenset(d3.keys()), set)
+        self.assertIsInstance(frozenset(d3.keys()) & d4.keys(), frozenset)
```
Comment (fgregg) [inline Lib/test/test_dictviews.py]: The current behavior is for 

``python
>>> frozenset([1, 2]) & {1: None}.keys()
{1}
```

But, 

``python
>>> frozenset([1, 2]) & {1, 2}
frozenset({1})
```

Should we prefer to keep the current behavior or make the intersection behave more like other intersections?

**Comment by fgregg** [comment]
The Python 3.7 behavior is

```python
>>> frozenset([{1, 2}) & {1: None}.keys()
{1}
>>> {1: None}.keys() &  frozenset([{1, 2})
{1}
```

But, 

```python
>>> frozenset({1, 2}) & {1}
frozenset({1})
>>> {1} & frozenset({1, 2})
{1}
```

Should we prefer to keep the current behavior or make the intersection behave more like other intersections?

**Comment by Mariatta** [comment]
Friendly reminder to use "Co-authored by" in the commit message in situations like this, where patch was originally started by someone else. Thanks.

[Devguide reference](https://devguide.python.org/pullrequest/?highlight=applying%20patch#converting-an-existing-patch-from-b-p-o-to-github)
[GitHub article](https://help.github.com/articles/creating-a-commit-with-multiple-authors/)

**Comment by fgregg** [comment]
Hi @Mariatta! 

I wasn't sure if that guidance applied here, because I didn't apply @dsu1995 patches and them make a new commit. Instead, I cherry-picked from his git branch. 

As such, David Su is already listed as the sole author of his commits.

If this is not the preferred option, I can redo this branch so that it does not have @dsu1995's commits and instead just has a commit where I applied his patch. In that commit I will list him as the Co-Author in the commit message.

Please advise.

File: Objects/dictobject.c [hunk 2]
```diff
@@ -3919,24 +3919,92 @@ dictviews_sub(PyObject* self, PyObject *other)
...
+    else {
+        dict_contains = dictitems_contains;
+    }
+
+    while ((key = PyIter_Next(it)) != NULL) {
+        rv = dict_contains((_PyDictViewObject *)self, key);
+        if (rv < 0)
+            goto error;
+        if (rv) {
+            if (PySet_Add(result, key)) {
```
Comment (ZackerySpytz) [inline Objects/dictobject.c:4245]: This `if` can be combined with the previous one.

File: Objects/dictobject.c [hunk 3]
```diff
@@ -3919,24 +3919,92 @@ dictviews_sub(PyObject* self, PyObject *other)
...
+        Py_DECREF(key);
+    }
+    Py_DECREF(it);
+    if (PyErr_Occurred()) {
+        Py_DECREF(result);
+        return NULL;
+    }
     return result;
+
+    error:
```
Comment (ZackerySpytz) [inline Objects/dictobject.c]: There is an extra level of indentation here (and under this label).

File: Objects/dictobject.c [hunk 4]
```diff
@@ -3919,24 +3919,92 @@ dictviews_sub(PyObject* self, PyObject *other)
...
+        Py_DECREF(it);
+        Py_DECREF(result);
+        Py_DECREF(key);
+        return NULL;
```
Comment (methane) [inline Objects/dictobject.c]: Dedent these 5 lines.

File: Objects/dictobject.c [hunk 5]
```diff
@@ -3919,24 +3919,92 @@ dictviews_sub(PyObject* self, PyObject *other)
...
 {
-    PyObject *result = PySet_New(self);
-    PyObject *tmp;
-    _Py_IDENTIFIER(intersection_update);
+    PyObject *result;
+    PyObject *it;
+    PyObject *key;
+    Py_ssize_t len_self;
+    int rv;
+    int (*dict_contains)(_PyDictViewObject *, PyObject *);
```
Comment (methane) [inline Objects/dictobject.c:4183]: Please define local variables when they are initialized.

For example

```
int rv = dict_contains((_PyDictViewObject *)self, key);
```

**Review by methane — CHANGES_REQUESTED** [review]

File: Misc/NEWS.d/next/Core and Builtins/2018-06-14-13-55-45.bpo-27575.mMYgzv.rst [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Improve speed of dictview intersection by directly using set intersection
+logic. Patch by David Hsu.
```
Comment (dsu1995) [inline Misc/NEWS.d/next/Core and Builtins/2018-06-14-13-55-45.bpo-27575.mMYgzv.rst]: typo: David Su

**Comment by csabella** [comment]
@fgregg, please address the code review requests.  Also, due to the code review requests, the 'Co-authored by:' seems appropriate.

File: Objects/dictobject.c (hunk 5 above)
Comment (rhettinger) [reply Objects/dictobject.c:4183]: It is also okay to list them all at the top.

File: Objects/dictobject.c (hunk 2 above)
Comment (rhettinger) [reply Objects/dictobject.c:4245]: This looks fine to me.

File: Objects/dictobject.c (hunk 3 above)
Comment (rhettinger) [reply Objects/dictobject.c]: Fixed.

File: Objects/dictobject.c (hunk 4 above)
Comment (rhettinger) [reply Objects/dictobject.c]: Done.

File: Misc/NEWS.d/next/Core and Builtins/2018-06-14-13-55-45.bpo-27575.mMYgzv.rst (hunk 6 above)
Comment (rhettinger) [reply Misc/NEWS.d/next/Core and Builtins/2018-06-14-13-55-45.bpo-27575.mMYgzv.rst]: Fixed.

**Comment by fgregg** [comment]
thanks for finishing this up, @rhettinger! And for the opportunity to contribute to CPython

**Comment by dsu1995** [comment]
Thank you @rhettinger! And thank you @fgregg for migrating my bpo patch to a Github PR and cleaning up the code!

**Pull request merged**